drivers/i2c/i2c_ll_stm32.c

```diff
@@ -103,6 +103,46 @@ static int i2c_stm32_init(struct device *dev)
 	__ASSERT_NO_MSG(clock);
 	clock_control_on(clock, (clock_control_subsys_t *) &cfg->pclken);
 
+#if defined(CONFIG_SOC_SERIES_STM32F3X) || defined(CONFIG_SOC_SERIES_STM32F0X)
+	/*
+	 * STM32F0/3 I2C's independent clock source doesn't support
+	 * APB1. Select I2C clock source according to Clock control
+	 * configuration.
+	 */
+
+	switch ((u32_t)cfg->i2c) {
+#ifdef CONFIG_I2C_1
+	case CONFIG_I2C_1_BASE_ADDRESS:
+		if (IS_ENABLED(CONFIG_STM32_PLL_SRC_HSI)) {
```

Comment:
I think this should be a Kconfig option.
User could chose I2C clock source to be either HSI or SYSCLK (with default SYSCLK).
If HSI, then you need to check if actually activated (in RCC registers, and not using Clock driver CONFIG_...)
if yes, select HSI
else trigger error
If SYSCLK (it should always be on, so it should be ok.)
 then select SYSCLK

Reply:
I agree with you that the driver should check the RCC registers and not the clock driver CONFIG.
I didn't find any LL function that checks if HSI is active. Is it ok if we read the register directly?
```READ_BIT(RCC->CR, RCC_CR_HSION) == (RCC_CR_HSION)```
I am skeptical about the option of selecting I2C clock from the Kconfig/dts file. It will be more confusing than the driver configuring it according to the configured system clock source.
If we decide to implement it using user input, we should to it for all families with independent I2C clock.

Reply:
> I didn't find any LL function that checks if HSI is active. Is it ok if we read the register directly?
READ_BIT(RCC->CR, RCC_CR_HSION) == (RCC_CR_HSION)

LL_RCC_HSI_IsReady is not strictly equivalent, but should provide the same service and will help you in code genericity

> I am skeptical about the option of selecting I2C clock from the Kconfig/dts file. It will be more confusing than the driver configuring it according to the configured system clock source.

Then, you'd only need to set to SYSCLK by default. Feeding I2C with HSI depending if HSI is activated or not make this uncertain and unflexible:
-Having HSI activated prevents you to have I2C feed by SYSCLK if you want to.
-You won't have same behavior depending on your HSI configuration (you could have requirements on HSI independant from I2C clocking).
For instance "Wakeup from Stop mode on address match" is available only when I2C clocked by HSI. What if you need HSI for other purpose but you don't want this feature.

> If we decide to implement it using user input, we should to it for all families with independent I2C clock.

Yes, this is only blocking for F3 right now, but we can say it should be available for all families. This could be done step by step though.
Other question would be : do we need to propose a clock selection API, as others IPs have the same constraint/options?

Reply:
I will set the I2C clock to SYSCLK for now. If there is a request/need I will come back to it.